contract_comm/currency/currency.go

```diff
@@ -94,97 +94,127 @@ const (
 )
 
 var (
-	cgExchangeRateNum = big.NewInt(1)
-	cgExchangeRateDen = big.NewInt(1)
-
 	medianRateFuncABI, _   = abi.JSON(strings.NewReader(medianRateABI))
 	balanceOfFuncABI, _    = abi.JSON(strings.NewReader(balanceOfABI))
 	getWhitelistFuncABI, _ = abi.JSON(strings.NewReader(getWhitelistABI))
 )
 
-type exchangeRate struct {
-	Numerator   *big.Int
-	Denominator *big.Int
+// ExchangeRate represent the exchangeRate for the pair (base, token)
+// Follows the equation: 1 base * ExchangeRate = X token
+type ExchangeRate struct {
+	numerator   *big.Int
+	denominator *big.Int
 }
 
-func ConvertToGold(val *big.Int, currencyFrom *common.Address) (*big.Int, error) {
-	if currencyFrom == nil {
-		return val, nil
+// NoopExchangeRate returns the noop rate
+// which results in a 1:1 conversion
+func NoopExchangeRate() *ExchangeRate { return nil }
+
+// MustNewExchangeRate creates an exchange rate, panic on error
+func MustNewExchangeRate(numerator *big.Int, denominator *big.Int) *ExchangeRate {
+	rate, err := NewExchangeRate(numerator, denominator)
+	if err != nil {
+		panic(err)
 	}
+	return rate
+}
 
-	celoGoldAddress, err := contract_comm.GetRegisteredAddress(params.GoldTokenRegistryId, nil, nil)
-	if err == errors.ErrSmartContractNotDeployed || err == errors.ErrRegistryContractNotDeployed {
-		log.Warn("Registry address lookup failed", "err", err)
-		return val, err
+// NewExchangeRate creates an exchange rate.
+// Requires numerator >=0 && denominator >= 0
+func NewExchangeRate(numerator *big.Int, denominator *big.Int) (*ExchangeRate, error) {
+	if numerator == nil || common.Big0.Cmp(numerator) >= 0 {
+		return nil, errors.ErrExchangeRateZero
 	}
+	if denominator == nil || common.Big0.Cmp(denominator) >= 0 {
+		return nil, errors.ErrExchangeRateZero
+	}
+	return &ExchangeRate{numerator, denominator}, nil
+}
 
-	if *currencyFrom == celoGoldAddress {
-		// This function shouldn't really be called with the token's address, but if it is the value
-		// is correct, so return nil as the error
-		return val, nil
-	} else if err != nil {
-		log.Error(err.Error())
-		return val, err
+// ToBase converts from token to base
+func (er *ExchangeRate) ToBase(tokenAmount *big.Int) *big.Int {
+	// check noop rate (cGLD rate)
+	if er == nil {
```

Comment:
Despite the fact that is compiles and works; I much rather have a `type NoOpExchangeRate struct` which implements an exchange rate interface. Better yet would be returning an exchange rate of 1.

Reply:
Example showing how this works: https://play.golang.org/p/tQe3kCMVAKq

Reply:
I personally don't have any issue with this implementation, meaningful `nil` values are a nice feature of Go. 

One option that would simplify things a bit though would be to create `ExchangeRate.Numerator()` and `ExchangeRate.Denominator()` functions that return `common.Big1` when `er == nil`. It would result in unnecessary multiplications and divisions by one, but that's not a huge issue.